packages/docusaurus/src/commands/writeHeadingIds.ts

```diff
@@ -38,77 +38,60 @@ function addHeadingId(
 
   const headingText = line.slice(headingLevel).trimEnd();
   const headingHashes = line.slice(0, headingLevel);
-  const slug = slugger
-    .slug(unwrapMarkdownLinks(headingText).trim(), {maintainCase})
-    .replace(/^-+/, '')
-    .replace(/-+$/, '');
+  const slug = slugger.slug(unwrapMarkdownLinks(headingText).trim(), {
+    maintainCase,
+  });
 
   return `${headingHashes}${headingText} {#${slug}}`;
 }
 
-export function transformMarkdownHeadingLine(
-  line: string,
-  slugger: Slugger,
+export function transformMarkdownContent(
+  content: string,
   options: Options = {maintainCase: false, overwrite: false},
 ): string {
   const {maintainCase = false, overwrite = false} = options;
-  if (!line.startsWith('#')) {
-    throw new Error(`Line is not a Markdown heading: ${line}.`);
-  }
-
-  const parsedHeading = parseMarkdownHeadingId(line);
+  const lines = content.split('\n');
+  const slugger = createSlugger();
 
-  // Do not process if id is already there
-  if (parsedHeading.id && !overwrite) {
-    return line;
+  // If we can't overwrite existing slugs, make sure other headings don't
+  // generate colliding slugs by first marking these slugs as occupied
+  if (!overwrite) {
+    lines.forEach((line) => {
+      const parsedHeading = parseMarkdownHeadingId(line);
+      if (parsedHeading.id) {
+        slugger.slug(parsedHeading.id);
+      }
+    });
   }
```

Comment:
Most of this is refactor and removing excess layers of abstraction, and this double-pass slugification is the only fix I made

Reply:
👍 nice trick

also wanted to use something similar unfortunately afaik the slugger does not offer an API to register forbidden slugs or something 😅